File: dotnet_csharp/HeroQuest.Test/HeroQuestAcceptanceTest.cs
namespace CodingDojo.Test;

using System.Text;

public class HeroQuestAcceptanceTest
{
    [Fact]
    public Task FullScenario()
    {
        HeroQuest.Output = new StringBuilder();

        QuestData questData = new()
        {
            PlayerName = "Conan",
            PlayerHealth = 100,
            PlayerStrength = 7,
            PlayerMagic = 15,
            PlayerCraftingSkill = 12,
            ItemName = "Healing Potion",
            ItemKind = "Health",
            ItemPower = 20,
            EnemyName = "Goblin Warlord",
            EnemyPower = 12
        };

        HeroQuest.Output.AppendLine("=== QUEST BEGINNING ===\n");
        string result = HeroQuest.PlayerToString(questData.PlayerName, questData.PlayerHealth, questData.PlayerStrength,
            questData.PlayerMagic, questData.PlayerCraftingSkill);
        HeroQuest.Output.AppendLine(result);

        result = HeroQuest.ItemToString(questData.ItemName, questData.ItemKind, questData.ItemPower);
        HeroQuest.Output.AppendLine(result);

        HeroQuest.Output.AppendLine("--- Exploring the dungeon... ---\n");
        questData.PlayerHealth = HeroQuest.PlayerFallsDown(questData.PlayerStrength, questData.PlayerHealth);
        result = HeroQuest.PlayerToString(questData.PlayerName, questData.PlayerHealth, questData.PlayerStrength,
            questData.PlayerMagic, questData.PlayerCraftingSkill);
        HeroQuest.Output.AppendLine(result);

        HeroQuest.Output.AppendLine("--- Using the healing item ---\n");
        (questData.PlayerHealth, questData.PlayerStrength, questData.PlayerMagic) = HeroQuest.ItemApplyEffectToPlayer(
            questData.ItemName, questData.ItemKind, questData.ItemPower, questData.PlayerHealth,
            questData.PlayerStrength, questData.PlayerMagic);
        result = HeroQuest.PlayerToString(questData.PlayerName, questData.PlayerHealth, questData.PlayerStrength,
            questData.PlayerMagic, questData.PlayerCraftingSkill);
        HeroQuest.Output.AppendLine(result);

        result = HeroQuest.ItemToString(questData.ItemName, questData.ItemKind, questData.ItemPower);
        HeroQuest.Output.AppendLine(result);

        HeroQuest.Output.AppendLine("--- Item degradation from repeated use ---\n");
        (questData.ItemKind, questData.ItemPower) =
            HeroQuest.ItemReduceByUsage(questData.ItemKind, questData.ItemPower);
        result = HeroQuest.ItemToString(questData.ItemName, questData.ItemKind, questData.ItemPower);
        HeroQuest.Output.AppendLine(result);

        (questData.ItemKind, questData.ItemPower) =
            HeroQuest.ItemReduceByUsage(questData.ItemKind, questData.ItemPower);
        result = HeroQuest.ItemToString(questData.ItemName, questData.ItemKind, questData.ItemPower);
        HeroQuest.Output.AppendLine(result);

        HeroQuest.Output.AppendLine("--- Repairing the damaged item ---\n");
        questData.ItemPower = HeroQuest.ItemRepair(questData.PlayerCraftingSkill, questData.ItemPower);
        result = HeroQuest.ItemToString(questData.ItemName, questData.ItemKind, questData.ItemPower);
        HeroQuest.Output.AppendLine(result);

        HeroQuest.Output.AppendLine("=== ENEMY ENCOUNTER ===\n");
        result = HeroQuest.EnemyToString(questData.EnemyName, questData.EnemyPower);
        HeroQuest.Output.AppendLine(result);

        questData.PlayerHealth = HeroQuest.EnemyAttackPlayer(questData.EnemyName, questData.EnemyPower,
            questData.PlayerStrength, questData.PlayerHealth);
        result = HeroQuest.PlayerToString(questData.PlayerName, questData.PlayerHealth, questData.PlayerStrength,
            questData.PlayerMagic, questData.PlayerCraftingSkill);
        HeroQuest.Output.AppendLine(result);

        HeroQuest.Output.AppendLine("--- Player retaliates ---\n");
        questData.EnemyPower = HeroQuest.PlayerChallengeEnemy(questData.EnemyName, questData.PlayerStrength,
            questData.ItemPower, questData.EnemyPower);
        result = HeroQuest.EnemyToString(questData.EnemyName, questData.EnemyPower);
        HeroQuest.Output.AppendLine(result);

        return Verifier.Verify(HeroQuest.Output.ToString());
    }
}

File: Java_mutable-parameters/src/test/java/codingdojo/HeroQuestAcceptanceTest.java
package codingdojo;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

public class HeroQuestAcceptanceTest {

    @Test
    public void fullScenario() {
        HeroQuest.output.setLength(0);

        String playerName = "Conan";
        int[] playerHealth = new int[]{100};
        int[] playerStrength = new int[]{7};
        int[] playerMagic = new int[]{15};
        int playerCraftingSkill = 12;
        String itemName = "Healing Potion";
        String[] itemKind = new String[]{"Health"};
        int[] itemPower = new int[]{20};
        String enemyName = "Goblin Warlord";
        int[] enemyPower = new int[]{12};

        HeroQuest.output.append("=== QUEST BEGINNING ===\n\n");
        String result = HeroQuest.playerToString(playerName, playerHealth[0], playerStrength[0],
                playerMagic[0], playerCraftingSkill);
        HeroQuest.output.append(result).append("\n");

        result = HeroQuest.itemToString(itemName, itemKind[0], itemPower[0]);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.output.append("--- Exploring the dungeon... ---\n\n");

        HeroQuest.playerFallsDown(playerHealth, playerStrength);

        result = HeroQuest.playerToString(playerName, playerHealth[0], playerStrength[0],
                playerMagic[0], playerCraftingSkill);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.output.append("--- Using the healing item ---\n\n");
        HeroQuest.itemApplyEffectToPlayer(
                itemName, itemKind, itemPower[0], playerHealth,
                playerStrength, playerMagic);

        result = HeroQuest.playerToString(playerName, playerHealth[0], playerStrength[0],
                playerMagic[0], playerCraftingSkill);
        HeroQuest.output.append(result).append("\n");

        result = HeroQuest.itemToString(itemName, itemKind[0], itemPower[0]);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.output.append("--- Item degradation from repeated use ---\n\n");
        HeroQuest.itemReduceByUsage(itemKind, itemPower);
        result = HeroQuest.itemToString(itemName, itemKind[0], itemPower[0]);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.itemReduceByUsage(itemKind, itemPower);
        result = HeroQuest.itemToString(itemName, itemKind[0], itemPower[0]);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.output.append("--- Repairing the damaged item ---\n\n");
        HeroQuest.itemRepair(itemPower, playerCraftingSkill);
        result = HeroQuest.itemToString(itemName, itemKind[0], itemPower[0]);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.output.append("=== ENEMY ENCOUNTER ===\n\n");
        result = HeroQuest.enemyToString(enemyName, enemyPower[0]);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.enemyAttackPlayer(enemyName, enemyPower[0],
                playerStrength, playerHealth);

        result = HeroQuest.playerToString(playerName, playerHealth[0], playerStrength[0],
                playerMagic[0], playerCraftingSkill);
        HeroQuest.output.append(result).append("\n");

        HeroQuest.output.append("--- Player retaliates ---\n\n");
        HeroQuest.playerChallengeEnemy(enemyName, playerStrength,
                itemPower, enemyPower);

        result = HeroQuest.enemyToString(enemyName, enemyPower[0]);
        HeroQuest.output.append(result).append("\n");

        Approvals.verify(HeroQuest.output.toString());
    }
}
